### DISCUSSION TITLE
Header file issue and IntegerDivideByZero Exception

### DISCUSSION THREAD

**Discussion by fscii** [pschatzmann/arduino-audio-tools#1568] [Q&A] [answered]
Board: ESP32 Dev
Arduino IDE: 2.3.2
Esp32 Core: 2.016
Example File: basic_a2dp_i2s.ino
arduino-audio-tools and ESP32-A2DP installed via 'git' method today.

First issue is header file inclusion - see comments in code for #include area - what causes this and how do I fix it?

After code, see exception report from AudioLogger::Debug.  Only changes to source was having to add AudioConfig.h for unexplainable reason since its already contained in AudioTools.h (???) and I2S settings for pins and Adafruit I2S format for MAX98357A board.

`#include "BluetoothA2DPSink.h"
#include "AudioTools.h"
#include "AudioConfig.h"  // wasn't here - had to add - won't compile without this - gives error I2SStream undefined -- ??!

BluetoothA2DPSink a2dp_sink;
I2SStream i2s;            // gives UNDEFINED error unless AudioConfig.h is included but AudioTools.h already includes AudioConfig.h ??!

// Write data to I2S
void read_data_stream(const uint8_t *data, uint32_t length) {
  i2s.write(data, length);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);  // was warning, setting to Debug

  // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);

  // Start Bluetooth Audio Receiver
  a2dp_sink.set_auto_reconnect(false);
  a2dp_sink.start("a2dp-i2s");

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);        // TX_MODE for Adafruit I2S
  
  cfg.pin_bck = 12;
  cfg.pin_ws = 14;
  cfg.pin_data = 27;  // try default pins in case mine are causing disaster!
  cfg.i2s_format = I2S_LSB_FORMAT;              // Adafruit I2S requirement

  cfg.sample_rate = a2dp_sink.sample_rate();
  cfg.channels = 2;
  cfg.bits_per_sample = 16;
  cfg.buffer_count = 8;
  cfg.buffer_size = 256;
  i2s.begin(cfg);
}

void loop() { delay(100); }`

Exception report:

`ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13964
load:0x40080400,len:3600
entry 0x400805f0
[D] I2SStream.h : 65 - bool audio_tools::I2SStream::begin(audio_tools::I2SConfig)
[D] AudioStreams.h : 33 - virtual void audio_tools::AudioStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 127 - in: sample_rate: 0 / channels: 2 / bits_per_sample: 16
[D] I2SESP32.h : 61 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32)
[D] I2SESP32.h : 180 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32, int, int)
[I] AudioTypes.h : 127 -  sample_rate: 0 / channels: 2 / bits_per_sample: 16
[I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 0
[I] I2SConfigESP32.h : 84 - bits per sample: 16
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_LSB_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 97 - buffer_count:8
[I] I2SConfigESP32.h : 98 - buffer_size:256
[I] I2SConfigESP32.h : 103 - pin_bck: 12
[I] I2SConfigESP32.h : 105 - pin_ws: 14
[I] I2SConfigESP32.h : 107 - pin_data: 27
[D] I2SESP32.h : 212 - i2s_driver_install
Guru Meditation Error: Core  1 panic'ed (IntegerDivideByZero). Exception was unhandled.

Core  1 register dump:
PC      : 0x400d8994  PS      : 0x00060730  A0      : 0x800d9700  A1      : 0x3ffcb7f0  
A2      : 0x00000000  A3      : 0x00000000  A4      : 0x3ffdfa84  A5      : 0x00000000  
A6      : 0x3ffdfa84  A7      : 0x3ffc4088  A8      : 0x00000100  A9      : 0x00000002  
A10     : 0x00000000  A11     : 0x00000001  A12     : 0x00000000  A13     : 0x3ffc4088  
A14     : 0xffffffbf  A15     : 0x00000008  SAR     : 0x0000001d  EXCCAUSE: 0x00000006  
EXCVADDR: 0x00000000  LBEG    : 0x400915ac  LEND    : 0x400915b7  LCOUNT  : 0x00000000  

Backtrace: 0x400d8991:0x3ffcb7f0 0x400d96fd:0x3ffcb840 0x400d30ee:0x3ffcb880 0x400d32b9:0x3ffcb900 0x400d3529:0x3ffcb970 0x400d3efd:0x3ffcb9e0 0x400d6d86:0x3ffcbaa0

ELF file SHA256: d8221aa412343b94
Rebooting...

Board Settings See attachment

![Screenshot 2024-05-21 181423](https://github.com/pschatzmann/arduino-audio-tools/assets/25595547/4ba2f314-c793-4eea-8f0f-1418b4ce7cce)

**Comment by fscii** [comment]
Update: - do this:

'#include "AudioTools.h"
#include "BluetoothA2DPSink.h"'

not this
'#include "BluetoothA2DPSink.h"
#include "AudioTools.h'

and suddenly include problem with requiring AudioConfig.h goes away. 

But exception remains.

**Reply by pschatzmann** [reply]
I changed the warning message to #  warning "AudioTools library is not included first or installed"

**Comment by pschatzmann** [comment, marked as answer]
Where do you have the example from that does the import of A2DP first ?

The crash comes from the fact that the sample rate is not defined: it must be 44100! 
The effective sample rate is only available after you have connected, so you need to either
- hardcode it to 44100 in your sketch like you did for the channels or bits_per_sample or 
- wait for the the connection status after calling start() with while(!a2dp_sink.is_connected()) delay(10);
- use the A2DP library  with the latest commit

**Comment by fscii** [comment]
The example comes from:
  arduino-audio-tools\examples\examples-communication\a2dp\basic-a2dp-i2s

the cfgcfg.sample_rate = 44100; and the while(!a2dp_sink.is_connected()) delay(10); both worked. I will stick with the delay so i2s can determine sample rate from a2dp.  I also decided to do the install via git so git pull can keep updates going.

I do get one warning even tho everything works, just for your information...

E (461) esp_core_dump_fl�K�ɕ�dump data check failed:
Calculated checksum='24acdc26'
Image checksum='ffffffff'

Everything works well.  Thank you for this library and your help.  Sending coffee!

**Reply by pschatzmann** [reply]
I corrected the basic-a2dp-i2s
